### PR TITLE
Source specific params

### PR DESCRIPTION
**Pull request by jpa198**
Add ability to make source specific parameters.
Added kwargs to perturbation functions so old versions remain compatible with the workflow, even if updates mean additional arguments are passed to perturbation scripts

### CODE REVIEW

File: srf_generation/source_parameter_generation/uncertainties/versions/V19p8.py [hunk 1]
```diff
@@ -20,7 +20,9 @@
 
 
 def generate_source_params(
-    sources_line: Union[GCMT_Source, NHM_Source]
+    sources_line: Union[GCMT_Source, NHM_Source],
+    additional_source_parameters: Dict[str, Any],
+    **kwargs,
```
Comment (shot0829) [inline srf_generation/source_parameter_generation/uncertainties/versions/V19p8.py]: probably a good idea to add a logger for precocious, so user can track what parameters are being used and what is not. ( dont wanna burn hundred thousands of core-hours just to realize the parameters are not being used)

File: srf_generation/source_parameter_generation/gcmt_to_realisation.py [hunk 2]
```diff
@@ -47,23 +47,17 @@
...
     parser = argparse.ArgumentParser()
 
-    parser.add_argument("version", type=str)
     parser.add_argument("fault_selection_file", type=abspath)
     parser.add_argument("gcmt_file", type=abspath)
+    parser.add_argument("--version", type=str, default="unperturbated")
     parser.add_argument("-n", "--n_processes", default=1, type=int)
     parser.add_argument("-c", "--cybershake_root", type=abspath, default=abspath("."))
     parser.add_argument("-a", "--aggregate_file", type=abspath)
+    parser.add_argument("--source_parameters", type=str, nargs="+")
```
Comment (jasonmotha) [inline srf_generation/source_parameter_generation/gcmt_to_realisation.py]: Help txt and nargs 2, specify multiple

File: srf_generation/source_parameter_generation/gcmt_to_realisation.py [hunk 3]
```diff
@@ -225,5 +255,40 @@ def main():
...
+    gcmt_lines,
+    perturbation_function,
+    primary_logger,
+):
+    messages = []
+    for i in range(len(faults)):
+        gcmt_data = gcmt_lines[i]
+
+        additional_source_specific_data = {}
+        print(gcmt_data[0])
```
Comment (jasonmotha) [inline srf_generation/source_parameter_generation/gcmt_to_realisation.py]: debug prints

File: srf_generation/source_parameter_generation/gcmt_to_realisation.py [hunk 4]
```diff
@@ -45,15 +45,28 @@
...
     parser.add_argument("-c", "--cybershake_root", type=abspath, default=abspath("."))
     parser.add_argument("-a", "--aggregate_file", type=abspath)
-    parser.add_argument("--source_parameters", type=str, nargs="+")
+    parser.add_argument(
+        "--source_parameter",
+        type=str,
+        nargs=2,
+        action="append",
+        metavar=("name", "filepath"),
+        help="Values to be passed to be added to each realisation, with one value per fault. "
```
Comment (jasonmotha) [inline srf_generation/source_parameter_generation/gcmt_to_realisation.py]: Include that you can specify this multiple times.

File: srf_generation/source_parameter_generation/gcmt_to_realisation.py [hunk 5]
```diff
@@ -213,15 +222,14 @@ def main():
 
     additional_source_parameters = pd.DataFrame()
 
-    if args.source_parameters is not None:
-        for i in range(len(args.source_parameters) // 2):
-            param, f_name = args.source_parameters[2 * i : 2 * i + 2]
+    if args.source_parameter is not None:
+        for param_name, filepath in args.source_parameter:
```
Comment (jasonmotha) [inline srf_generation/source_parameter_generation/gcmt_to_realisation.py]: That is so much nicer.

File: srf_generation/source_parameter_generation/gcmt_to_realisation.py [hunk 6]
```diff
@@ -52,9 +52,22 @@ def load_args(primary_logger: Logger):
...
     parser.add_argument("gcmt_file", type=abspath)
     parser.add_argument("--version", type=str, default="unperturbated")
-    parser.add_argument("-n", "--n_processes", default=1, type=int)
+    parser.add_argument(
+        "-n",
+        "--n_processes",
+        default=1,
+        type=int,
+        help="Number of processes to generate realisations with. "
+        "Setting this higher than the number of faults will not have any additional effect.",
```
Comment (jasonmotha) [inline srf_generation/source_parameter_generation/gcmt_to_realisation.py]: This is capped to the number of faults to process.

File: srf_generation/source_parameter_generation/gcmt_to_realisation.py [hunk 7]
```diff
@@ -80,15 +95,14 @@ def load_args(primary_logger: Logger):
...
-            else:
-                args.source_parameter[i][1] = filepath
+    for i, (param_name, filepath) in enumerate(args.source_parameter):
+        if not isfile(filepath):
+            errors.append(
+                f"The file {filepath} given for parameter "
+                f"{param_name} does not exist"
+            )
+        else:
+            args.source_parameter[i][1] = filepath
```
Comment (jasonmotha) [inline srf_generation/source_parameter_generation/gcmt_to_realisation.py]: Ummm, isn't that already filepath?

Comment (jpa198) [reply srf_generation/source_parameter_generation/gcmt_to_realisation.py]: Good catch, not anymore

File: srf_generation/source_parameter_generation/gcmt_to_realisation.py [hunk 8]
```diff
@@ -96,6 +96,7 @@ def load_args(primary_logger: Logger):
         )
 
     for i, (param_name, filepath) in enumerate(args.source_parameter):
+        filepath = abspath(filepath)
```
Comment (jasonmotha) [inline srf_generation/source_parameter_generation/gcmt_to_realisation.py]: https://stackoverflow.com/questions/55257955/getting-full-path-from-the-relative-path-with-argparse/55258475#55258475

this is slightly nicer than modifying argparse. I generally don't expect changes to argparse values after parse args. I'm not quite sure how it works for nargs=2 but probably a tuple coming in.

File: srf_generation/source_parameter_generation/gcmt_to_realisation.py [hunk 9]
```diff
@@ -45,25 +45,47 @@
...
+    parser.add_argument("--version", type=str, default="unperturbated")
+    parser.add_argument(
+        "-n",
+        "--n_processes",
+        default=1,
+        type=int,
+        help="Number of processes to generate realisations with. "
+        "Setting this higher than the number of faults will not have any additional effect.",
+    )
     parser.add_argument("-c", "--cybershake_root", type=abspath, default=abspath("."))
```
Comment (jpa198) [inline srf_generation/source_parameter_generation/gcmt_to_realisation.py]: @jasonmotha Dude...

File: srf_generation/source_parameter_generation/gcmt_to_realisation.py (hunk 8 above)
Comment (jpa198) [reply srf_generation/source_parameter_generation/gcmt_to_realisation.py]: The result returned from load_args is the same

**Pull request merged**